### PR TITLE
add near-api-js cookbook

### PR DESCRIPTION
**Pull request by thisisjoshford**
Adds example common use cases for `near-api-js`.
- access key rotation
- calculating gas
- account creation w/o NEAR Wallet
- read state w/o an account
- sign and send batch transactions
- query recent transaction information w/o indexer

### CODE REVIEW

File: docs/api/naj-cookbook.md [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+const args = {
+  text: "Howdy!",
+};
+
+const config = {
+  networkId: "testnet",
+  nodeUrl: "https://rpc.testnet.near.org",
+  walletUrl: "https://wallet.testnet.near.org",
+  helperUrl: "https://helper.testnet.near.org",
+};
```
Comment (mehtaphysical) [inline docs/api/naj-cookbook.md:430]:
```suggestion
const config = {
  keyStore,
  networkId: "testnet",
  nodeUrl: "https://rpc.testnet.near.org",
};
```

File: docs/api/naj-cookbook.md [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+
+calculateGas(CONTRACT_ID, METHOD_NAME, args, "0");
+
+async function calculateGas(contractId, methodName, args, depositAmount) {
+  const credentialsPath = path.join(homedir, CREDENTIALS_DIR);
+  const keyStore = new keyStores.UnencryptedFileSystemKeyStore(credentialsPath);
```
Comment (mehtaphysical) [inline docs/api/naj-cookbook.md:436]: Move these above the function like the other examples

File: docs/api/naj-cookbook.md [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+  const near = await connect({ ...config, keyStore });
```
Comment (mehtaphysical) [inline docs/api/naj-cookbook.md:437]: Make keyStore part of config:

```suggestion
  const near = await connect(config);
```

File: docs/api/naj-cookbook.md [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+  "https://rpc.testnet.near.org"
+);
+
+getState();
+
+async function getState() {
+  const rawResult = await provider.query(
+    `call/guest-book.testnet/getMessages`,  // function / contract name / contract method
+    "AQ4"                                   // Base 58 of '{}'
+  );
```
Comment (mehtaphysical) [inline docs/api/naj-cookbook.md:499]: Use the newer style RPC call:

```suggestion
  const rawResult = await provider.query({
    request_type: "call_function",
    account_id: 'guest-book.testnet',
    method_name: 'getMessages',
    args_base64: 'e30=',
    finality: 'optimistic'
  }); 
```

File: docs/api/naj-cookbook.md [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+getState();
+
+async function getState() {
+  const rawResult = await provider.query(
+    `call/guest-book.testnet/getMessages`,  // function / contract name / contract method
+    "AQ4"                                   // Base 58 of '{}'
+  ); 
+  // format result
+  const res = JSON.parse(
+    rawResult.result.map((x) => String.fromCharCode(x)).join("")
```
Comment (mehtaphysical) [inline docs/api/naj-cookbook.md:502]:
```suggestion
    Buffer.from(rawResult.result).toString()
```

File: docs/api/naj-cookbook.md [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+  for (let i = 0; i < result.receipts_outcome.length; i++) {
+    gasBurnt.push(result.receipts_outcome[i].outcome.gas_burnt);
+    tokensBurnt.push(
+      utils.format.formatNearAmount(
+        result.receipts_outcome[i].outcome.tokens_burnt
+      )
+    );
+  }
+  const totalGasBurned = gasBurnt.reduce((acc, cur) => acc + cur, 0);
+  const totalTokensBurned = tokensBurnt.reduce((acc, cur) => acc + cur, 0);
```
Comment (mehtaphysical) [inline docs/api/naj-cookbook.md:463]:
```suggestion
  const account = await near.account(ACCOUNT_ID);
  const result = await account.functionCall({
    contractId,
    methodName,
    args,
    gas: "300000000000000",
    attachedDeposit: utils.format.parseNearAmount(depositAmount),
  });

  const { totalGasBurned, totalGasBurned } = result.receipts_outcome.reduce(
    (acc, receipt) => {
      acc.totalGasBurned += receipt.outcome.gas_burnt;
      acc.totalTokensBurned += utils.format.formatNearAmount(receipt.outcome.tokens_burnt)
    },
    {
      totalGasBurned: result.transaction_outcome.outcome.gas_burnt,
      totalTokensBurned: utils.format.formatNearAmount(
        result.transaction_outcome.outcome.tokens_burnt
      ),
    }
  );

```

File: docs/api/naj-cookbook.md [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+const homedir = require("os").homedir();
+
+const CREDENTIALS_DIR = ".near-credentials";
+const CONTRACT_ID = "example.testnet";
+const AUTHORIZED_CONTRACT = "example-contract.testnet";
+const METHODS = ["example_method"];
+const ALLOWANCE = "2500000000000";
+
+const credentialsPath = path.join(homedir, CREDENTIALS_DIR);
+const keyStore = new keyStores.UnencryptedFileSystemKeyStore(credentialsPath);
```
[the remark this answers is not in the document]
Comment (thisisjoshford) [reply docs/api/naj-cookbook.md:193]: Wasn't 100% sure where it best belongs. Thanks for the suggestion, I'll move it :)

File: docs/api/naj-cookbook.md (hunk 6 above)
Comment (thisisjoshford) [reply docs/api/naj-cookbook.md:463]: Awesome! Nice refactor! I think on line 447 you meant `totalGasBurned, totalTokensBurned`. Updating this now

File: docs/api/naj-cookbook.md (hunk 1 above)
Comment (mehtaphysical) [reply docs/api/naj-cookbook.md]: missed this one @thisisjoshford

**Pull request merged**